File: src/main/java/net/yogstation/yogbot/listeners/commands/GetNotesCommand.java
package net.yogstation.yogbot.listeners.commands;

import net.yogstation.yogbot.DatabaseManager;
import net.yogstation.yogbot.config.DiscordConfig;
import net.yogstation.yogbot.permissions.PermissionsManager;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public abstract class GetNotesCommand extends PermissionsCommand {
	protected final DatabaseManager database;
	
	public GetNotesCommand(DiscordConfig discordConfig, PermissionsManager permissions, DatabaseManager database) {
		super(discordConfig, permissions);
		this.database = database;
	}
	
	protected List<String> getNotes(String ckey, boolean showAdmin) {
		try (Connection connection = database.getConnection();
			 PreparedStatement notesStmt = connection.prepareStatement(String.format(
				 "SELECT timestamp, text, adminckey FROM `%s` WHERE `targetckey` = ? AND `type`= \"note\" AND deleted = 0 AND (expire_timestamp > NOW() OR expire_timestamp IS NULL) AND `secret` = 0 ORDER BY `timestamp`",
				 database.prefix("messages")
			 ))
		) {
			notesStmt.setString(1, ckey);
			ResultSet notesResult = notesStmt.executeQuery();

			final List<String> messages = new ArrayList<>();
			StringBuilder notesString = new StringBuilder("Notes for ").append(ckey).append("\n");
			while (notesResult.next()) {
				String nextNote = String.format(
					"```%s\t%s%s```",
					notesResult.getDate("timestamp").toString(),
					notesResult.getString("text"),
					showAdmin ? String.format("   %s", notesResult.getString("adminckey")) : ""
				);
				if (notesString.length() + nextNote.length() > 2000) {
					messages.add(notesString.toString());
					notesString.setLength(0);
				}
				notesString.append(nextNote);
			}
			messages.add(notesString.toString());
			return messages;
		} catch (SQLException e) {
			LOGGER.error("Error getting notes", e);
			return List.of("A SQL Error has occurred");
		}
	}
}

File: src/main/java/net/yogstation/yogbot/listeners/commands/GetNotesCommand.kt
package net.yogstation.yogbot.listeners.commands

import net.yogstation.yogbot.DatabaseManager
import net.yogstation.yogbot.config.DiscordConfig
import net.yogstation.yogbot.permissions.PermissionsManager
import java.sql.SQLException

/**
 * Gets the notes of a player, to be used either for a player getting their own notes, or an admin getting someone else's
 */
abstract class GetNotesCommand(
	discordConfig: DiscordConfig,
	permissions: PermissionsManager,
	protected val database: DatabaseManager
) : PermissionsCommand(
	discordConfig, permissions
) {
	/**
	 * Gets the notes as a list
	 * @param ckey The ckey to get notes for
	 * @param showAdmin Should the admin's name be included
	 */
	protected fun getNotes(ckey: String, showAdmin: Boolean): List<String> {
		try {
			database.byondDbConnection.use { connection ->
				connection.prepareStatement(
					String.format(
						"SELECT timestamp, text, adminckey FROM `%s` WHERE `targetckey` = ? AND `type`= \"note\" AND deleted = 0 AND (expire_timestamp > NOW() OR expire_timestamp IS NULL) AND `secret` = 0 ORDER BY `timestamp`",
						database.prefix("messages")
					)
				).use { notesStmt ->
					notesStmt.setString(1, ckey)
					val notesResult = notesStmt.executeQuery()
					val messages: MutableList<String> = ArrayList()
					val notesString = StringBuilder("Notes for ").append(ckey).append("\n")
					while (notesResult.next()) {
						val nextNote = "```${notesResult.getDate("timestamp")}\t${notesResult.getString("text")}${
							if (showAdmin) "   ${
								notesResult.getString(
									"adminckey"
								)
							}" else ""
						}```"
						if (notesString.length + nextNote.length > 2000) {
							messages.add(notesString.toString())
							notesString.setLength(0)
						}
						notesString.append(nextNote)
					}
					messages.add(notesString.toString())
					return messages
				}
			}
		} catch (e: SQLException) {
			logger.error("Error getting notes", e)
			return listOf("A SQL Error has occurred")
		}
	}
}
